File: src/cpp/BERDoubleLongUnsigned.cpp
#include "BERCodes.h"
#include "BERDoubleLongUnsigned.h"
#include "BERException/BERExceptions.h"
#include "BasicBuffer.h"
#include <stdint.h>
#include <list>
#include <sstream>
using namespace std;

namespace BER
{
	uint32_t BERDoubleLongUnsigned::Size(uint32_t *value)
	{
		return 5;
	}

	BasicBuffer* BERDoubleLongUnsigned::Encode(uint32_t *value)
	{
		BasicBuffer *result = new BasicBuffer(Size(value));
		result->PushBack(BER::DoubleLongUnsigned);
		for (int8_t i = 3; i >= 0; i--)
		{
			result->PushBack(((*value) >> (i * 8)) & 0xFF);
		}
		return result;
	}

	uint32_t* BERDoubleLongUnsigned::Decode(BasicBuffer &buffer, uint32_t &index)
	{
		if (buffer.Get(index) != BER::DoubleLongUnsigned)
		{
			throw BERTypeIncorrectTag();
		}
		index++;
		uint32_t *temp = new uint32_t(0);
		*temp |= buffer.Get(index++) << 24;
		*temp |= buffer.Get(index++) << 16;
		*temp |= buffer.Get(index++) <<  8;
		*temp |= buffer.Get(index++);
		return temp;
	}

	BERDoubleLongUnsigned::~BERDoubleLongUnsigned()
	{
		DELETE(this->Value);
	}

	string BERDoubleLongUnsigned::toString()
	{
		ostringstream oss;
		oss << "{BERDoubleLongUnsigned: " << (uint32_t)(*Value) << "}" ;
		return oss.str();
	}
}

File: src/bercs/BERDoubleLongUnsigned.cs
﻿using System;
using System.Collections.Generic;

using BER.Exceptions;

namespace BER
{
    public class BERDoubleLongUnsigned : BaseBERType<UInt32>
    {
        #region Overrides of BaseBERType<uint>
		//! Method that encodes the value of this BERDoubleLongUnsigned into a Byte List
		//! \return A Byte List containing the encoded raw bytes of this BERDoubleLongUnsigned	
        public override List<byte> Encode(uint value)
        {
            var result = new List<byte> {(Byte) BERCodes.DoubleLongUnsigned};
            for (SByte i = 3; i >= 0; i--)
            {
                result.Add((Byte) ((value >> (i * 8)) & 0xFF));
            }
            return result;
        }
		//! Method that decodes a List of Bytes into this BERDoubleLongUnsigned object
		//! \return true The decoding process was succesfull
		//! \return true The decoding process was not succesfull
		//! \throws BERTypeDecodeIncorrectTag if an incompatible tag was found
        public override uint Decode(List<byte> buffer)
        {
            if (buffer[0] != (int) BERCodes.DoubleLongUnsigned)
            {
                throw new BERTypeDecodeIncorrectTag();
            }
            var result = (UInt32) ((buffer[1] << 24) | (buffer[2] << 16) | (buffer[3] << 8) | buffer[4]);
            buffer.RemoveRange(0, 5);
            return result;
        }

        #endregion
		//! Method that provides a String representation of this BERDoubleLongUnsigned
		//! \return The string representation of this BERDoubleLongUnsigned
        public override string AsString()
        {
            return Value.ToString();
        }

    }
}
